dotnet/roslyn
# Implement simple LSP run tests

First part of https://github.com/dotnet/vscode-csharp/issues/5719

Implements the server side code to handle the run tests command and the codelens for the same.

Client side PR - https://github.com/dotnet/vscode-csharp/pull/5897

The general runtests approach is as follows
* The client does as little as possible - it calls the server whenever it sees a run tests requests and passes whatever range it was given.
  * We save the files to ensure the contents are up to date on the server when we build.
  * The server streams in messages and progress which the client writes to the test output window and updates the progress bar (cancellable).
* The server then implements the actual run tests command
  * Shell out to the dotnet CLI to build the project.  This is required to ensure the the output dll is relatively up to date with the request snapshot.  We cannot guarantee that the server snapshot will 100% match the build output, but we do our best.  O# does something similar.
  * Runs test discovery
    * First we search the request snapshot for possible test methods in a given range.  This essentially just looks for methods with a matching attribute name.  We don't desire 100% accuracy and neither does O#.  We'll leave 100% correctness to devkit.
    * Shell out to the vstest console to run test discovery on the project output dll.
    * Figure out which discovered tests are the ones we want to run from the potential methods.
  * Run the discovered tests
    * Shells out to the vstest console to run the discovered test cases.  Fairly straightforward, but we need to prettify the output in the console.
* For codelens we use the same potential test method discovery to tag test methods and test classes.  The actual codelens is just a command that has the test method range to run.



 ![run_tests](https://github.com/dotnet/roslyn/assets/5749229/02cad2be-ac27-4fdb-880b-be7f6dbaae6f)


Followups in next PRs
1.  run settings
4.  finish localization
5.  debug commands
